packages/connect-conformance/src/callback-client.ts

```diff
@@ -0,0 +1,250 @@
+        }
+      },
+      (err) => {
+        if (err !== undefined) {
+          error = ConnectError.from(err);
+          // We can't distinguish between headers and trailers here, so we just
+          // add the metadata to both.
+          //
+          // But if the headers are already set, we don't need to overwrite them.
+          resHeaders =
+            resHeaders.length === 0
+              ? convertToProtoHeaders(error.metadata)
+              : resHeaders;
+          resTrailers = convertToProtoHeaders(error.metadata);
+        }
+        resolve(
+          new ClientResponseResult({
+            responseHeaders: resHeaders,
+            responseTrailers: resTrailers,
+            payloads: payloads,
+            error: convertToProtoError(error),
+          }),
+        );
+      },
+      {
+        headers: reqHeader,
+        onHeader(headers) {
+          resHeaders = convertToProtoHeaders(headers);
+        },
+        onTrailer(trailers) {
+          resTrailers = convertToProtoHeaders(trailers);
+        },
+      },
+    );
+  });
+}
+
+async function unimplemented(
+  client: ConformanceClient,
+  req: ClientCompatRequest,
+) {
+  const msg = req.requestMessages[0];
+  const unReq = new UnimplementedRequest();
+  if (!msg.unpackTo(unReq)) {
+    throw new Error("Could not unpack request message to unary request");
+  }
+  const reqHeader = new Headers();
+  appendProtoHeaders(reqHeader, req.requestHeaders);
+  let error: ConnectError | undefined = undefined;
+  let resHeaders: ConformanceHeader[] = [];
+  let resTrailers: ConformanceHeader[] = [];
+
+  return new Promise<ClientResponseResult>((resolve) => {
+    client.unimplemented(
+      unReq,
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      (err, _) => {
+        if (err !== undefined) {
+          error = ConnectError.from(err);
+          // We can't distinguish between headers and trailers here, so we just
+          // add the metadata to both.
+          //
+          // But if the headers are already set, we don't need to overwrite them.
+          resHeaders =
+            resHeaders.length === 0
+              ? convertToProtoHeaders(error.metadata)
+              : resHeaders;
+          resTrailers = convertToProtoHeaders(error.metadata);
+        }
+        resolve(
+          new ClientResponseResult({
+            responseHeaders: resHeaders,
+            responseTrailers: resTrailers,
+            error: convertToProtoError(error),
+          }),
+        );
+      },
+      {
+        headers: reqHeader,
+        onHeader(headers) {
+          resHeaders = convertToProtoHeaders(headers);
+        },
+        onTrailer(trailers) {
+          resTrailers = convertToProtoHeaders(trailers);
+        },
+      },
+    );
+  });
+}
+
+export function invoke(
+  transport: Transport,
+  req: ClientCompatRequest,
+): Promise<ClientResponseResult> {
+  const client = createCallbackClient(ConformanceService, transport);
```

Comment:
I forgot to mention this earlier: Can we move the export at the top of the file to make it easier to read? 

Now that we have two versions, I'd prefer if we name the function `invokeWithCallbackClient` / `invokeWithPromiseClient` to be more self-describing.

Same for promise-client.ts